### ISSUE TITLE
"Cannot initialize filter fcidecomp: File not found"

### ISSUE THREAD

**Issue by texadactyl** [silx-kit/hdf5plugin#112]
This message started appearing in the last several days.

Envirtonment
O/S: Ubuntu 20.10, up-to-date
hdf5plugin version: 2.3.2
h5py version: 3.2.1
Python 3.8.6
All of my packages are managed under pip 20.1.1.

I just recently observed the subject message which seems to originate from the hdf5plugin. 
```
import h5py
import hdf5plugin
print('Right after importing hdf5plugin')
```
gets this stdout:
```
Cannot initialize filter fcidecomp: File not found
Right after importing hdf5plugin
```
Switch the order of the 2 imports, I see the same thing.

**Comment by texadactyl** [comment]
When I did this:
```
pip uninstall -y h5py hdf5plugin
pip install  h5py==2.10.0 hdf5plugin==2.3.2
```
and then ran the above tiny program, this message does not appear.

**Comment by texadactyl** [comment]
So, it looks like an h5py issue.

I'll leave this open so you'll see it but please feel free to mark it as "not my issue"  (-:

**Comment by vasole** [comment]
Not sure where is the problem. Have you tried other h5py 3.x versions?

**Comment by texadactyl** [comment]
Not yet.  2.10 is not alarming to the scientists that I work with.  (-:

**Comment by texadactyl** [comment]
They were worried about their data getting compromised.  2.10 is a version that was on one of the data pipeline servers so I just tried it out.

**Comment by t20100** [comment]
Hi,

I could not reproduce this with Ubuntu 20.04, Python 3.8.5, pip 20.1.1, hdf5plugin 2.3.2, h5py 3.2.1.

**Comment by takluyver** [comment]
It's not obvious to me how this could be an h5py issue, as the code involved in making that error message doesn't seem to touch h5py - it's just finding a file that should be part of the hdf5plugin package:

https://github.com/silx-kit/hdf5plugin/blob/3df5c33cb99add78ef5367a0738276aac033ad3b/hdf5plugin/__init__.py#L301-L306

**Comment by t20100** [comment]
The `fcidecomp` filter requires C++11, so if C++11 is not available, it is not built and you would get this message.
To avoid this message in this case, we would need to store build options (related issue: #96).

Did you installed `hdf5plugin` from source or from wheels?

We provide 2 Linux wheels for `hdf5plugin`:
- `manylinux1`: C++11 not available and so no `fcidecomp` filter.
- `manylinux2014`: With C++11 and so with `fcidecomp` filter.

Maybe `pip` picked the `manylinux1` wheel?

**Comment by takluyver** [comment]
It could have been installed by an older version of pip which didn't recognise `manylinux2014`.

If some of the filters are meant to be optional, maybe it makes sense to drop the warning on import and give an informative error when someone tries to use FciDecomp, e.g. "The fcidecomp filter is not available, as this build of hdf5plugin could not use C++11. You may need to reinstall hdf5plugin with a recent version of pip, or rebuild it with a newer compiler."

**Comment by t20100** [comment]
Yes, we need to store build time options to use them at runtime to improve this.

**Comment by takluyver** [comment]
If that's complicated (I see #96 has been open a few months), you could cheat a bit by treating the presence of fcidecomp as a record of the build configuration. :slightly_smiling_face:

**Comment by t20100** [comment]
When the `hdf5plugin.FciDecomp` class is used without the filter being available, we can provide an error message, but for decompression, I think there is nothing we can do.

Regarding #96, it just stayed low on the priority list so far.

**Comment by takluyver** [comment]
Oh yes, good point, I'd forgotten that you can decompress without ever explicitly specifying the filter. :flushed:

**Comment by texadactyl** [comment]
Ironically, my scientific organization does not use the fcidecomp filter.  Hence, the surprise at seeing it mentioned at all in a message.  We use the Bitshuffle filter with lz4.  Have been for several years.  The subject message just turned up in the lats 7-10 days.

We install only with pip, either directly or as a part of someone's requirements.txt file.

**Comment by t20100** [comment]
Maybe due to upgrading `hdf5plugin` recently?

Fcidecomp was added in May last year (https://github.com/silx-kit/hdf5plugin/releases/tag/v2.2.0).

Anyway, we'll fix this warning, thanks for reporting it.

**Comment by vasole** [comment]
> We use the Bitshuffle filter with lz4. Have been for several years. The subject message just turned up in the last 7-10 days.

Then as far as you do not get an error related to Bitshuffle+lz4 everything will be fine for you.

We also take care to be able to build hdf5plugin also from source. Therefore by using ``pip install hdf5plugin --no-binary`` you could get a version better optimized for your platform/compiler.

**Comment by texadactyl** [comment]
@vasole Thanks!

**Comment by vasole** [comment]
What puzzles me in this issue is that installing `pip install  h5py==2.10.0 hdf5plugin==2.3.2` does not trigger the message at the OP end but the same combination with `h5py==3.2.1` seems to do it. If the hdf5plugin mode is already installed and not giving an error, `pip install h5py==3.2.1` should be fine.

> For performance reasons, should we get both h5py and hdf5plugin with the --no-binary option?

No, tt is not so straightforward to build h5py from source.

**Comment by texadactyl** [comment]
So just hdf5plugin --no-binary

**Comment by t20100** [comment]
@vasole, yes, regarding this issue, the behaviour should be the same no matter of the `h5py` version.
It should only depend on the kind of wheel that was installed.

@texadactyl could you run just `pip install --upgrade h5py==3.2.1` (i.e., not changing `hdf5plugin`) on an environment whereyou don't have the error with `h5py==2.10.0` to check whether there is still no error (i.e., the issue was the kind of wheel installed) or not (and there is something else going on)?

**Comment by texadactyl** [comment]
We will keep an eye out for these messages.  I have alerted folks at Seti BL - dozens of servers.  Maybe my deadstart procedure will resolve any future reoccurences.

**Comment by t20100** [comment]
Thanks for the testing!

We've fixed the error message in case fcidecomp filter was not built, so next release will fix that anyway.

**Comment by texadactyl** [comment]
Which culprit?  My money is on "pip".  It's happened before.  (:
You're welcome!
